### PR TITLE
[KianYee] #459 enhance file info function

### PR DESCRIPTION
**Pull request by TSS-LauKY**
Requirement for https://github.com/oceanprotocol/market/issues/459 & https://github.com/oceanprotocol/market/pull/479.

Changes proposed in this PR:

- if arg is DID type, get provider endpoint from DDO
- if arg is string type, get provider endpoint from getFileinfoEndpoint()

### CODE REVIEW

**Comment by alexcos20** [comment]
cool.  can you add some unit tests as well ?

**Comment by TSS-LauKY** [comment]
I added a negative test, since we have cover positive test.

File: src/provider/Provider.ts [hunk 1]
```diff
@@ -494,4 +492,23 @@ export class Provider extends Instantiable {
       return false
     }
   }
+
+  private async getServiceEndpoint(did: string): Promise<string> {
```
Comment (mihaisc) [inline src/provider/Provider.ts:496]: Why do we need this? We already have `getServiceEndpoints` that returns all the endpoints. This function seems to return the compute service if found, and if not return the access service. But what happens if you have both? This is kind of hardcoding. A better approach would be to add `serviceIndex` as a function argument and then check that service. The one who uses this function needs to decide what service he wants to use.

Comment (TSS-LauKY) [reply src/provider/Provider.ts:496]: Yes, we can change to serviceIndex. 

getServiceEndpoints will return the endpoint base on config, there is no issue if we want to check fileinfo by passing url.
If we need to check the fileinfo for existing DID, and the provider from DDO is not the same with endpoint from config (wallet connection), we are checking the wrong endpoint.

Comment (mihaisc) [reply src/provider/Provider.ts:496]: > getServiceEndpoints will return the endpoint base on config, there is no issue if we want to check fileinfo by passing url.

Yes, my bad.

I'm a little confused about the argument name `url`, isn't that a did? Or can you give me some examples of what url are you refering to?

Comment (TSS-LauKY) [reply src/provider/Provider.ts:496]: the argument url can be use as:

- file url: 

- mostly we are using for data publish

image

- DID:

- provider connectivity check once DDO is crated. in this case we need to get the provider from DDO instead for config.

image

Comment (TSS-LauKY) [reply src/provider/Provider.ts:496]: I'm not sure is it better to split this function into 2

- filelinfo: pass url as argument and check the connectivity base on endpoint from config and return File response
- isFileconsumable: pass did as argument and check the connectivity base on endpoint from DDO and return Boolean

Comment (mihaisc) [reply src/provider/Provider.ts:496]: sounds good, less confusing and more usability

Comment (TSS-LauKY) [reply src/provider/Provider.ts:496]: I'm going to create isFileConsumable then.
Is it better have 2 argument: did and serviceIndex?

File: test/integration/Marketplaceflow.test.ts [hunk 2]
```diff
@@ -400,10 +400,12 @@ describe('Marketplace flow', () => {
   })
 
   it('Alice should check if her asset has valid url(s)', async () => {
```
Comment (mihaisc) [inline test/integration/Marketplaceflow.test.ts:402]: update test case to something like '.... check if her asset is consumable' and the it's good to go

**Review by mihaisc — APPROVED** [review]
looks good

**Comment by kremalicious** [comment]
worth noting that on `market` we do not use `ocean.provider.fileinfo()` but have our own helper method to hit the provider directly:
https://github.com/oceanprotocol/market/blob/main/src/utils/provider.ts

Using `ocean.provider.fileinfo()` will only work when user has a wallet connected which is a restriction because of ocean.js architecture we work around on market.

**Pull request merged**